File: 04/2.hs
main = do
    txt <- readFile "input.txt"
    let (min, max) = break (== '-') txt
    let [min', max'] = map read [min, tail max]
    let n = length $ filter (==True) $ map matches [min'..max']
    print n

matches :: Int -> Bool
matches n = decreasing n && hasdouble n
    where
        decreasing n =
            let s = show n
                xs = zip [0..] s
            in all (\(i,c) -> i == 0 || c >= s !! (i - 1)) xs
        hasdouble n = 
            let s = show n
                xs = zip [1..] $ tail s
                ch i
                    | i < 0 = ' '
                    | i >= length s = ' '
                    | otherwise = s !! i
            in any (\(i,c) -> c == ch (i - 1) && c /= ch (i - 2) && c /= ch (i + 1)) xs

File: 04/2.js
const fs = require("fs")
const [min, max] = fs.readFileSync("input.txt").toString().split("-").map(Number)

let n = 0

for (let i = min; i <= max; i++) {
    let s = i.toString().split("")
    if (
        s.every((c, idx) => idx === 0 || c >= s[idx - 1]) &&
        s.some((c, idx) => c === s[idx - 1] && c !== s[idx - 2] && c !== s[idx + 1])
    ) {
        n++
    }
}

console.log(n)
